src/main/scala/chiseltest/formal/AddResetAssumptionPass.scala

```diff
@@ -0,0 +1,123 @@
+      ir.NoInfo,
+      namespace.newName("_resetCount"),
+      resetCountType,
+      clock,
+      reset = ir.Reference(preset).copy(flow = SourceFlow),
+      init = Utils.getGroundZero(resetCountType)
+    )
+    val resetCountSource = ir.Reference(resetCount).copy(flow = SourceFlow)
+
+    // we are in the reset phase iff the counter is less than the reset length
+    val lessThanCycles = ir.DoPrim(
+      PrimOps.Lt,
+      List(resetCountSource, ir.UIntLiteral(resetLength, resetCountType.width)),
+      List(),
+      Utils.BoolType
+    )
+    val resetPhase = ir.DefNode(ir.NoInfo, namespace.newName("_resetPhase"), lessThanCycles)
+    val resetPhaseRef = ir.Reference(resetPhase).copy(flow = SourceFlow)
+
+    // we increment the counter as long as we are in the reset phase, after that it just retains its value
+    val nextValue = Utils.mux(resetPhaseRef, plusOne(resetCountSource), resetCountSource)
+    val nextCon = ir.Connect(ir.NoInfo, ir.Reference(resetCount).copy(flow = SinkFlow), nextValue)
+
+    // add assumption that reset is active
+    val resetActive = ir.Verification(
+      ir.Formal.Assume,
+      ir.NoInfo,
+      clock,
+      pred = reset,
+      en = resetPhaseRef,
+      msg = ir.StringLit(""),
+      name = namespace.newName("_resetActive")
+    )
+
+    // collect all our statements and add them to the main module
+    val stmts = Seq(resetCount, resetPhase, nextCon, resetActive)
+    val instrumented = main.copy(ports = main.ports :+ preset, body = ir.Block(main.body +: stmts))
+
+    // substitute instrumented main and add annotations
+    val otherMods = state.circuit.modules.filterNot(_.name == state.circuit.main)
+    state.copy(
+      circuit = state.circuit.copy(modules = instrumented +: otherMods),
+      annotations = presetAnno +: state.annotations
+    )
+  }
+
+  def getResetLength(annos: AnnotationSeq): Int = {
+    annos.collect { case ResetOption(n) => n }.distinct.toList match {
+      case List()    => 1 // default
+      case List(one) => one
+      case more =>
+        throw new RuntimeException(s"Received multiple disagreeing reset options! " + more.mkString(", "))
+    }
+  }
+
+  private def resetIsPreset(main: String, annos: AnnotationSeq): Boolean = {
+    annos.collectFirst {
+      case PresetAnnotation(target) if target.circuit == main && target.module == main && target.ref == "reset" => true
+    }.getOrElse(false)
+  }
+
+  private def findClockAndReset(m: ir.Module): (ir.Reference, ir.Reference) = {
+    val clock = m.ports
+      .find(_.name == "clock")
+      .getOrElse(
+        throw new RuntimeException(s"[${m.name}] Expected module to have a port named clock!")
+      )
+    val reset = m.ports
+      .find(_.name == "reset")
+      .getOrElse(
+        throw new RuntimeException(s"[${m.name}] Expected module to have a port named reset!")
+      )
+    (ir.Reference(clock).copy(flow = SourceFlow), ir.Reference(reset).copy(flow = SourceFlow))
+  }
```

Comment:
This limit the dut to be `Module` with clock and reset, In the future, maybe we can add another pass to find all real clock and reset for a `RawModule`.